sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -968,6 +970,10 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
                 new FileResult<>(writer.getOutputFile(), shard, window, c.pane(), entry.getKey())));
         closeWriterInBackground(writer);
       }
+
+      // Ensure that the past closes happen before returning and after we started the closes
```

Comment:
nit: I find the happen before/after confusing since they may happen whenever we are just blocking on them completing here

Maybe
Block on completing the past closes before returning. We do so after starting the current closes in the background
so that they can happen in parallel.

Reply:
Done, thanks for making this clearer.